flow/src/main/java/flow/State.java

```diff
@@ -22,24 +22,41 @@
 import android.support.annotation.Nullable;
 import android.util.SparseArray;
 import android.view.View;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static flow.Preconditions.checkNotNull;
 
 public class State {
+  private static final String VIEW_STATE_IDS = "VIEW_STATE_IDS";
+  private static final String BUNDLE = "BUNDLE";
+  private static final String VIEW_STATE_PREFIX = "VIEW_STATE_";
+  private static final String KEY = "KEY";
+
   /** Creates a State instance that has no state and is effectively immutable. */
   @NonNull public static State empty(@NonNull final Object key) {
     return new EmptyState(key);
   }
 
   @NonNull static State fromBundle(@NonNull Bundle savedState, @NonNull KeyParceler parceler) {
-    Object key = parceler.toKey(savedState.getParcelable("KEY"));
+    Object key = parceler.toKey(savedState.getParcelable(KEY));
     State state = new State(key);
-    state.viewState = savedState.getSparseParcelableArray("VIEW_STATE");
-    state.bundle = savedState.getBundle("BUNDLE");
+    int[] viewIds = checkNotNull(savedState.getIntArray(VIEW_STATE_IDS), "Null view state ids?");
+    for (int viewId : viewIds) {
+      SparseArray<Parcelable> viewState =
+          savedState.getSparseParcelableArray(VIEW_STATE_PREFIX + viewId);
+      if (viewState != null) {
```

Comment:
Might not be: the view could have `saveEnabled="false"` and there's this check in the `toBundle` method:
```
      SparseArray<Parcelable> viewState = entry.getValue();
      if (viewState.size() > 0) {
        outState.putSparseParcelableArray(VIEW_STATE_PREFIX + viewId, viewState);
      }
```